### ISSUE TITLE
[Bug] RuntimeError: CUDA error: CUBLAS_STATUS_EXECUTION_FAILED when calling `cublasSgemm( handle, opa, opb, m, n, k, &alpha, a, lda, b, ldb, &beta, c, ldc)` when training tacotron2

### ISSUE THREAD

**Issue by blackmamba1122** [coqui-ai/TTS#1517]
## 🐛 Description

i got the UserWarning when i try to training tacotron2 with ljspeech. the warning is as below:
```
/media/DATA-2/TTS/coqui/TTS/TTS/tts/models/tacotron2.py:341: UserWarning: __floordiv__ is deprecated, and its behavior will change in a future version of pytorch. It currently rounds toward 0 (like the 'trunc' function NOT 'floor'). This results in incorrect rounding for negative values. To keep the current behavior, use torch.div(a, b, rounding_mode='trunc'), or for actual floor division, use torch.div(a, b, rounding_mode='floor').
  alignment_lengths = mel_lengths // self.decoder.r
```

and after epoch 8/1000  in step   1886/3243 i got error like this:

```
RuntimeError: CUDA error: CUBLAS_STATUS_EXECUTION_FAILED when calling `cublasSgemm( handle, opa, opb, m, n, k, &alpha, a, lda, b, ldb, &beta, c, ldc)`
```
Is that the cause of the error? how to handle it?

I will display the error in its entirety

```
   --> STEP: 1886/3243 -- GLOBAL_STEP: 27830
     | > decoder_loss: 5.84848  (6.74442)
     | > postnet_loss: 6.73053  (7.62051)
     | > stopnet_loss: 0.29443  (0.42176)
     | > decoder_coarse_loss: 6.32641  (7.27556)
     | > decoder_ddc_loss: 0.00271  (0.00433)
     | > ga_loss: 0.00282  (0.00469)
     | > decoder_diff_spec_loss: 1.54124  (1.60230)
     | > postnet_diff_spec_loss: 3.17193  (3.19602)
     | > decoder_ssim_loss: 0.91339  (0.91384)
     | > postnet_ssim_loss: 0.92837  (0.92731)
     | > loss: 6.67432  (7.51629)
     | > align_error: 0.98659  (0.97846)
     | > grad_norm: 3.66964  (5.65185)
     | > current_lr: 0.00000 
     | > step_time: 0.72660  (0.52215)
     | > loader_time: 0.00150  (0.00552)

 ! Run is kept in /media/DATA-2/TTS/coqui/TTS/run-April-19-2022_01+40PM-0cf3265a
Traceback (most recent call last):
  File "/media/DATA-2/TTS/coqui/tts_coqui/lib/python3.8/site-packages/trainer/trainer.py", line 1461, in fit
    self._fit()
  File "/media/DATA-2/TTS/coqui/tts_coqui/lib/python3.8/site-packages/trainer/trainer.py", line 1445, in _fit
    self.train_epoch()
  File "/media/DATA-2/TTS/coqui/tts_coqui/lib/python3.8/site-packages/trainer/trainer.py", line 1224, in train_epoch
    _, _ = self.train_step(batch, batch_num_steps, cur_step, loader_start_time)
  File "/media/DATA-2/TTS/coqui/tts_coqui/lib/python3.8/site-packages/trainer/trainer.py", line 1057, in train_step
    outputs, loss_dict_new, step_time = self._optimize(
  File "/media/DATA-2/TTS/coqui/tts_coqui/lib/python3.8/site-packages/trainer/trainer.py", line 946, in _optimize
    outputs, loss_dict = self._model_train_step(batch, model, criterion)
  File "/media/DATA-2/TTS/coqui/tts_coqui/lib/python3.8/site-packages/trainer/trainer.py", line 902, in _model_train_step
    return model.train_step(*input_args)
  File "/media/DATA-2/TTS/coqui/TTS/TTS/tts/models/tacotron2.py", line 344, in train_step
    outputs = self.forward(text_input, text_lengths, mel_input, mel_lengths, aux_input)
  File "/media/DATA-2/TTS/coqui/TTS/TTS/tts/models/tacotron2.py", line 206, in forward
    decoder_outputs, alignments, stop_tokens = self.decoder(encoder_outputs, mel_specs, input_mask)
  File "/media/DATA-2/TTS/coqui/tts_coqui/lib/python3.8/site-packages/torch/nn/modules/module.py", line 1110, in _call_impl
    return forward_call(*input, **kwargs)
  File "/media/DATA-2/TTS/coqui/TTS/TTS/tts/layers/tacotron/tacotron2.py", line 321, in forward
    decoder_output, attention_weights, stop_token = self.decode(memory)
  File "/media/DATA-2/TTS/coqui/TTS/TTS/tts/layers/tacotron/tacotron2.py", line 284, in decode
    decoder_output = self.linear_projection(decoder_hidden_context)
  File "/media/DATA-2/TTS/coqui/tts_coqui/lib/python3.8/site-packages/torch/nn/modules/module.py", line 1110, in _call_impl
    return forward_call(*input, **kwargs)
  File "/media/DATA-2/TTS/coqui/TTS/TTS/tts/layers/tacotron/common_layers.py", line 25, in forward
    return self.linear_layer(x)
  File "/media/DATA-2/TTS/coqui/tts_coqui/lib/python3.8/site-packages/torch/nn/modules/module.py", line 1110, in _call_impl
    return forward_call(*input, **kwargs)
  File "/media/DATA-2/TTS/coqui/tts_coqui/lib/python3.8/site-packages/torch/nn/modules/linear.py", line 103, in forward
    return F.linear(input, self.weight, self.bias)
RuntimeError: CUDA error: CUBLAS_STATUS_EXECUTION_FAILED when calling `cublasSgemm( handle, opa, opb, m, n, k, &alpha, a, lda, b, ldb, &beta, c, ldc)`
```

### Environment

{
    "CUDA": {
        "GPU": [
            "NVIDIA GeForce GTX 1660 Ti"
        ],
        "available": true,
        "version": "10.2"
    },
    "Packages": {
        "PyTorch_debug": false,
        "PyTorch_version": "1.11.0+cu102",
        "TTS": "0.6.1",
        "numpy": "1.19.5"
    },
    "System": {
        "OS": "Linux",
        "architecture": [
            "64bit",
            "ELF"
        ],
        "processor": "x86_64",
        "python": "3.8.0",
        "version": "#118~18.04.1-Ubuntu SMP Thu Mar 3 13:53:15 UTC 2022"
    }
}

### Additional context

I've reduced the batch size, currently I'm using batch size = 4. Is this also caused by reducing the batch size? But if it's not reduced it will be OOM.

**Comment by WeberJulian** [comment]
Hey, could you please try with the latest CUDA 11 version?

**Comment by blackmamba1122** [comment]
> Hey, could you please try with the latest CUDA 11 version?

Okey, I'll try with the CUDA 11.3 version

**Comment by blackmamba1122** [comment]
> > Hey, could you please try with the latest CUDA 11 version?
> 
> Okey, I'll try with the CUDA 11.3 version

I had trying, but still error

```
   --> STEP: 267/3243 -- GLOBAL_STEP: 3510
     | > decoder_loss: 33.00844  (33.08039)
     | > postnet_loss: 35.04267  (35.08060)
     | > stopnet_loss: 0.81244  (0.85546)
     | > decoder_coarse_loss: 32.94590  (33.04457)
     | > decoder_ddc_loss: 0.00334  (0.00528)
     | > ga_loss: 0.00708  (0.01021)
     | > decoder_diff_spec_loss: 0.41570  (0.43186)
     | > postnet_diff_spec_loss: 4.45244  (4.44248)
     | > decoder_ssim_loss: 0.99999  (0.99990)
     | > postnet_ssim_loss: 0.99990  (0.99950)
     | > loss: 27.81493  (27.92766)
     | > align_error: 0.97789  (0.96763)
     | > grad_norm: 4.54514  (5.27340)
     | > current_lr: 0.00000 
     | > step_time: 0.30480  (0.25091)
     | > loader_time: 0.00150  (0.03358)

 ! Run is kept in /media/DATA-2/TTS/coqui/TTS/run-April-22-2022_01+20PM-0cf3265a
Traceback (most recent call last):
  File "/media/DATA-2/TTS/coqui/tts_coqui/lib/python3.8/site-packages/trainer/trainer.py", line 1485, in fit
    self._fit()
  File "/media/DATA-2/TTS/coqui/tts_coqui/lib/python3.8/site-packages/trainer/trainer.py", line 1469, in _fit
    self.train_epoch()
  File "/media/DATA-2/TTS/coqui/tts_coqui/lib/python3.8/site-packages/trainer/trainer.py", line 1248, in train_epoch
    _, _ = self.train_step(batch, batch_num_steps, cur_step, loader_start_time)
  File "/media/DATA-2/TTS/coqui/tts_coqui/lib/python3.8/site-packages/trainer/trainer.py", line 1081, in train_step
    outputs, loss_dict_new, step_time = self._optimize(
  File "/media/DATA-2/TTS/coqui/tts_coqui/lib/python3.8/site-packages/trainer/trainer.py", line 1018, in _optimize
    loss_dict["loss"].backward()
  File "/media/DATA-2/TTS/coqui/tts_coqui/lib/python3.8/site-packages/torch/_tensor.py", line 363, in backward
    torch.autograd.backward(self, gradient, retain_graph, create_graph, inputs=inputs)
  File "/media/DATA-2/TTS/coqui/tts_coqui/lib/python3.8/site-packages/torch/autograd/__init__.py", line 173, in backward
    Variable._execution_engine.run_backward(  # Calls into the C++ engine to run the backward pass
RuntimeError: CUDA error: CUBLAS_STATUS_EXECUTION_FAILED when calling `cublasSgemm( handle, opa, opb, m, n, k, &alpha, a, lda, b, ldb, &beta, c, ldc)`

```
## Environment
{
    "CUDA": {
        "GPU": [
            "NVIDIA GeForce GTX 1660 Ti"
        ],
        "available": true,
        "version": "11.3"
    },
    "Packages": {
        "PyTorch_debug": false,
        "PyTorch_version": "1.11.0+cu113",
        "TTS": "0.6.1",
        "numpy": "1.19.5"
    },
    "System": {
        "OS": "Linux",
        "architecture": [
            "64bit",
            "ELF"
        ],
        "processor": "x86_64",
        "python": "3.8.0",
        "version": "#123~18.04.1-Ubuntu SMP Fri Apr 8 09:48:52 UTC 2022"
    }
}

**Comment by erogol** [comment]
It is, in general, a sign of insufficient VRAM. You can try a GPU with a larger VRAM or reduce the batch size or limit the maximum audio length allowed in training.

**Comment by blackmamba1122** [comment]
> It is, in general, a sign of insufficient VRAM. You can try a GPU with a larger VRAM or reduce the batch size or limit the maximum audio length allowed in training.

Hi, i've try to reduce the batch size. Now I'm using 4 for batch size.

```
import os

from trainer import Trainer, TrainerArgs

from TTS.config.shared_configs import BaseAudioConfig
from TTS.tts.configs.shared_configs import BaseDatasetConfig
from TTS.tts.configs.tacotron2_config import Tacotron2Config
from TTS.tts.datasets import load_tts_samples
from TTS.tts.models.tacotron2 import Tacotron2
from TTS.tts.utils.speakers import SpeakerManager
from TTS.tts.utils.text.tokenizer import TTSTokenizer
from TTS.utils.audio import AudioProcessor

output_path = os.path.dirname(os.path.abspath(__file__))
dataset_config = BaseDatasetConfig(name="vctk", meta_file_train="", path=os.path.join(output_path, "../VCTK/"))

audio_config = BaseAudioConfig(
    sample_rate=22050,
    resample=True,  # Resample to 22050 Hz. It slows down training. Use `TTS/bin/resample.py` to pre-resample and set this False for faster training.
    do_trim_silence=True,
    trim_db=23.0,
    signal_norm=False,
    mel_fmin=0.0,
    mel_fmax=8000,
    spec_gain=1.0,
    log_func="np.log",
    preemphasis=0.0,
)

config = Tacotron2Config(  # This is the config that is saved for the future use
    audio=audio_config,
    batch_size=4,
    eval_batch_size=4,
    num_loader_workers=2,
    num_eval_loader_workers=2,
    run_eval=True,
    test_delay_epochs=-1,
    r=2,
    # gradual_training=[[0, 6, 48], [10000, 4, 32], [50000, 3, 32], [100000, 2, 32]],
    double_decoder_consistency=False,
    epochs=1000,
    text_cleaner="phoneme_cleaners",
    use_phonemes=True,
    phoneme_language="en-us",
    phoneme_cache_path=os.path.join(output_path, "phoneme_cache"),
    print_step=10,
    print_eval=False,
    mixed_precision=True,
    min_text_len=0,
    #max_text_len=500,
    max_text_len=400,
    min_audio_len=0,
    #max_audio_len=44000 * 10,
    max_audio_len=40000 * 10,
    output_path=output_path,
    datasets=[dataset_config],
    use_speaker_embedding=True,  # set this to enable multi-sepeaker training
    decoder_ssim_alpha=0.0,  # disable ssim losses that causes NaN for some runs.
    postnet_ssim_alpha=0.0,
    postnet_diff_spec_alpha=0.0,
    decoder_diff_spec_alpha=0.0,
    attention_norm="softmax",
    optimizer="Adam",
    lr_scheduler=None,
    lr=3e-5,
)
```
But I got the error again

```
   --> STEP: 520/525 -- GLOBAL_STEP: 520
     | > decoder_loss: 2.31593  (6.74750)
     | > postnet_loss: 2.41630  (6.36996)
     | > stopnet_loss: 0.21022  (0.61206)
     | > ga_loss: 0.00368  (0.00860)
     | > loss: 1.41170  (3.93444)
     | > align_error: 0.97678  (0.94576)
     | > amp_scaler: 32768.00000  (35918.76923)
     | > grad_norm: 1.17756  (4.16678)
     | > current_lr: 0.00003 
     | > step_time: 2.93800  (1.06941)
     | > loader_time: 0.00180  (0.00159)

 ! Run is kept in /media/botika/DATA-2/TTS/TTS_Coqui/TTS/run-June-07-2022_10+20AM-68cef28a
Traceback (most recent call last):
  File "/media/botika/DATA-2/TTS/TTS_Coqui/coqui_env/lib/python3.7/site-packages/trainer/trainer.py", line 1492, in fit
    self._fit()
  File "/media/botika/DATA-2/TTS/TTS_Coqui/coqui_env/lib/python3.7/site-packages/trainer/trainer.py", line 1476, in _fit
    self.train_epoch()
  File "/media/botika/DATA-2/TTS/TTS_Coqui/coqui_env/lib/python3.7/site-packages/trainer/trainer.py", line 1255, in train_epoch
    _, _ = self.train_step(batch, batch_num_steps, cur_step, loader_start_time)
  File "/media/botika/DATA-2/TTS/TTS_Coqui/coqui_env/lib/python3.7/site-packages/trainer/trainer.py", line 1097, in train_step
    num_optimizers=len(self.optimizer) if isinstance(self.optimizer, list) else 1,
  File "/media/botika/DATA-2/TTS/TTS_Coqui/coqui_env/lib/python3.7/site-packages/trainer/trainer.py", line 1008, in _optimize
    scaler.scale(loss_dict["loss"]).backward()
  File "/media/botika/DATA-2/TTS/TTS_Coqui/coqui_env/lib/python3.7/site-packages/torch/_tensor.py", line 363, in backward
    torch.autograd.backward(self, gradient, retain_graph, create_graph, inputs=inputs)
  File "/media/botika/DATA-2/TTS/TTS_Coqui/coqui_env/lib/python3.7/site-packages/torch/autograd/__init__.py", line 175, in backward
    allow_unreachable=True, accumulate_grad=True)  # Calls into the C++ engine to run the backward pass
RuntimeError: CUDA error: CUBLAS_STATUS_EXECUTION_FAILED when calling `cublasGemmEx( handle, opa, opb, m, n, k, &falpha, a, CUDA_R_16F, lda, b, CUDA_R_16F, ldb, &fbeta, c, CUDA_R_16F, ldc, CUDA_R_32F, CUBLAS_GEMM_DFALT_TENSOR_OP)`
```

Environment:
```
{
    "CUDA": {
        "GPU": [
            "NVIDIA GeForce GTX 1660 Ti"
        ],
        "available": true,
        "version": "10.2"
    },
    "Packages": {
        "PyTorch_debug": false,
        "PyTorch_version": "1.11.0+cu102",
        "TTS": "0.6.2",
        "numpy": "1.21.6"
    },
    "System": {
        "OS": "Linux",
        "architecture": [
            "64bit",
            "ELF"
        ],
        "processor": "x86_64",
        "python": "3.7.13",
        "version": "#49~20.04.1-Ubuntu SMP Wed May 18 18:44:28 UTC 2022"
    }
}
```
I'm using 11.3 CUDA version if I running ```nvcc --version```

```
nvcc: NVIDIA (R) Cuda compiler driver
Copyright (c) 2005-2021 NVIDIA Corporation
Built on Mon_May__3_19:15:13_PDT_2021
Cuda compilation tools, release 11.3, V11.3.109
Build cuda_11.3.r11.3/compiler.29920130_0
````

**Comment by blackmamba1122** [comment]
This issue was resolved by adding memory

**Comment by blackmamba1122** [comment]
This issue was resolved by adding memory

**Comment by ch13d2am** [comment]
Hi @blackmamba1122,

I am facing the same issue.

Can you please give more detail on what you mean by "adding memory"? Did you use a different GPU with larger VRAM?

Thanks.

**Comment by blackmamba1122** [comment]
> Hi @blackmamba1122,
> 
> I am facing the same issue.
> 
> Can you please give more detail on what you mean by "adding memory"? Did you use a different GPU with larger VRAM?
> 
> Thanks.

Hi, sorry I just read your comment. Yes, before I used RAM: 16 GiB dual channel, and now I used 32 GiB dual channel. But it only works in batch_size=4.

**Comment by Fransie-dev** [comment]
I experienced this issue on Linux and I solved it by running

```bash
$ unset LD_LIBRARY_PATH
```

**Comment by aneet-k** [comment]
I got this error too. It's funny that re-executing the code solved the problem for me.

**Comment by RamatovInomjon** [comment]
> ```shell
> unset LD_LIBRARY_PATH
> ```

Saved my day! Thanks!

**Comment by ZixuWang** [comment]
I still got this error after trying 'unset LD_LIBRARY_PATH'. The thing for me is it can run on CPU. But not on cuda

**Comment by cyberkarim** [comment]
I'm still having this issue after running 'unset LD_LIBRARY_PATH'

**Comment by chungisaac** [comment]
> 
``` bash
unset LD_LIBRARY_PATH
```

Hoping to understand this issue deeper. For those who found success in this command, why does this command work for you?

**Comment by AndrewTKent** [comment]
> I experienced this issue on Linux and I solved it by running
> 
> ```shell
> $ unset LD_LIBRARY_PATH
> ```

I LOVE YOU

**Comment by luoyoucode** [comment]
> I experienced this issue on Linux and I solved it by running
> 
> ```shell
> $ unset LD_LIBRARY_PATH
> ```

I figured this out!!! THX! I don't understand it, but I was just in awe.

**Comment by Fatima-Naseem071** [comment]
> I still got this error after trying 'unset LD_LIBRARY_PATH'. The thing for me is it can run on CPU. But not on cuda

Same. Can someone help?

**Comment by jacoblambert** [comment]
Also fixed for me after `$ unset LD_LIBRARY_PATH` and would love to know why.

**Comment by iwannabewater** [comment]
> I experienced this issue on Linux and I solved it by running
>
> ```shell
> $ unset LD_LIBRARY_PATH
> ```

Thank you very much, this worked for me, but I specifically want to know why?

**Comment by Net-Maker** [comment]
I used `unset LD_LIBRARY_PATH` and it solved my problem. I also wonder why this can work. 
I need to switch CUDA versions frequently, so I wrote a script myself to modify CUDA_PATH and LD_LIBRARY_PATH, which is also the reason why I triggered this bug,  refer to this [link](https://discuss.pytorch.org/t/runtimeerror-cuda-error-cublas-status-invalid-value-when-calling-cublassgemm-handle-opa-opb-m-n-k-alpha-a-lda-b-ldb-beta-c-ldc/124544/35), I guess this error is caused by linking your system's LD_LIBRARY_PATH to the program, but what the program really needs is the dynamic link library in conda.
